### PR TITLE
Refine SKILL.md: principles, structure, examples

### PR DESCRIPTION
**Pull request by DmitrySharabin**
## Summary

Iterates on `SKILL.md` from a real session using it to refactor a downstream project's tests. Three themes: frontload principles, cut duplications, capture gotchas.

## Key changes

- **Restructure:** Best Practices + Common Mistakes moved up to right after Core Pattern, so principles frame the mechanics that follow.
- **Two foundational principles** atop Best Practices, as meta-rules that override specific guidance below on conflict:
  - *What abstraction makes the tests easier to understand for humans?*
  - *Write the tests first; scaffold around them.*
- **New Common Mistakes rows:** reading values via `this.arg.X` inside `run` (use the parameter list); `//` comments for test intent (use `description`, kept brief).
- **Tightened rows:** `expect: true` (smell, with yes/no-question-name exception); custom `run` per test (identical bodies vs unique imperative content).
- **New `data` example:** setup hook stashing a per-test fixture for `run`, with the hook-vs-`run` parameter-access asymmetry called out.
- **Docs reference:** top-of-file pointer to <https://htest.dev> + specific link to [`data` docs](https://htest.dev/define/#data).
- **Removed duplications:** `args` vs `arg` warning, "don't test the language", over-prescriptive `run (value)` row.

Preview: https://deploy-preview-130--h-test.netlify.app/skill.html

### CODE REVIEW

File: SKILL.md [hunk 1]
```diff
@@ -29,6 +31,33 @@ export default {
 };
 ```
 
+## Best Practices
+
+**Two principles guide everything below.** Come back to these whenever a specific rule seems wrong for your test:
+
+- **What abstraction makes the tests easier to understand for humans?** If a rule below would impose an abstraction that makes a test harder to read, the rule is wrong for that test. Declarativeness exists to remove repetitive noise — not to force inherently imperative code into a data shape.
+
+- **Write the tests first; scaffold around them.** Express each test as data (`name`, `arg`, `expect`) in whatever shape gives you maximum signal-to-noise on the intent. Write `run()`, `beforeEach()`, and other scaffolding *at the end*, derived from what the tests need. Starting with `run()` first shapes the tests around what `run()` can express, instead of letting the tests express the intent.
```
Comment (LeaVerou) [inline SKILL.md]: Maybe add something around "The goal is maximizing signal-to-noise ratio, not characters. DRY is not the endgame, it just often correlates."

But also `run()` can sometimes be legit part of the test, like in the ones we were discussing. 

Maybe instead of discouraging run() altogether, tell the agent to write the test both with and without a run() and evaluate what expresses intent better?

File: SKILL.md [hunk 2]
```diff
@@ -29,6 +31,33 @@ export default {
...
+2. **Structure mirrors API** — nest test groups to match your module's shape. One export = one top-level group. The hierarchy should let you locate any test by navigating the API.
+3. **Check the shape, not every byte** — use `check: { subset: true }` when you only care about specific properties. Tests that over-specify `expect` break for wrong reasons.
+4. **Each test = one unique branch** — don't test the same code path twice with different data. A passing duplicate isn't a safety net.
+5. **Don't test the language** — spread, object merge, array methods: those are JavaScript, not your code. Only test behavior your function owns.
+
+## Common Mistakes
+
+| Mistake                                               | Fix                                                                                          |
+| ----------------------------------------------------- | -------------------------------------------------------------------------------------------- |
+| Custom `run` in every test with identical bodies      | Extract shared `run` at parent. **But** if each test's `run` has unique imperative content (writes, sequences), per-test is correct — don't impose uniformity. |
```
Comment (LeaVerou) [inline SKILL.md]: I'd say with minor differences rather than identical

File: SKILL.md [hunk 3]
```diff
@@ -29,6 +31,33 @@ export default {
...
+| Reading values via `this.arg.X` / `this.args[0]` inside `run` | The parameter list IS where `arg`/`args` arrive. Name and unpack them however fits the case: `run ({ x })`, `run ([first, ...rest])`, `run (args) { let first = args[0]; }`, etc. Reserve `this` for `this.data` (lifecycle state) and instance props. |
+| Inline `//` comment explaining what the test verifies | Use the `description` field — that's its home as structured metadata next to the test. **Keep it brief** (a sentence, maybe two): add what the `name` can't carry — an issue reference, a subtle invariant, a non-obvious edge case. Don't write a paragraph and don't restate the name in prose. Reserve comments for short notes about mechanics. |
+| Setup in `beforeAll` but used in `arg` expressions    | Move to module top level — `arg`/`expect` evaluate at import time, before hooks              |
+| `new Instance()` in both `arg` and `expect`           | Share one instance variable                                                                  |
```
Comment (LeaVerou) [inline SKILL.md]: Example? (for me, not the docs)

File: SKILL.md [hunk 4]
```diff
@@ -122,8 +151,8 @@ export default {
 
 | Property                   | Description                                                                                    |
 | -------------------------- | ---------------------------------------------------------------------------------------------- |
-| `beforeEach` / `afterEach` | Run before/after each test. Inherited. `afterEach` runs even if the test throws. Sync or async |
-| `beforeAll` / `afterAll`   | Run before/after all tests in the group where defined. **Not inherited**                       |
+| `beforeEach` / `afterEach` | Run before/after each test. Receive no parameters — access state via `this.data`, `this.arg`, etc. Inherited. `afterEach` runs even if the test throws. Sync or async |
```
Comment (LeaVerou) [inline SKILL.md:154]: I wonder if we _should_ be passing parameters to them!

**Review by LeaVerou — CHANGES_REQUESTED** [review]
See comments, but otherwise this looks great!

File: SKILL.md (hunk 3 above)
Comment (DmitrySharabin) [reply SKILL.md]: The row's pointing at hTest's default check using `===` at leaf level — two structurally-equal-but-distinct instances fail. Short example:

```js
// ❌ Two `new Map()` calls — different instances, fails
{ arg: new Map([["k", "v"]]), expect: { result: new Map([["k", "v"]]) } }

// ✅ Share one instance
let map = new Map([["k", "v"]]);
{ arg: map, expect: { result: map } }
```

Same pattern is spelled out under [Reference Equality for Instances](https://github.com/htest-dev/htest/blob/refine-skill-md/SKILL.md#reference-equality-for-instances) later in the same file.

File: SKILL.md (hunk 4 above)
Comment (DmitrySharabin) [reply SKILL.md:154]: They probably need `arg`/`args` in most cases. And yes, I think it's useful. Everything else can be accessed via `this`.

File: SKILL.md (hunk 1 above)
Comment (DmitrySharabin) [reply SKILL.md]: Updated. Take a look, please.

File: SKILL.md (hunk 4 above)
Comment (LeaVerou) [reply SKILL.md:154]: Please file an issue so we can track this?

**Review by LeaVerou — APPROVED** [review]
